pkg/cloudprovider/openstack.go

```diff
@@ -237,45 +219,126 @@ func (o *OpenStack) AssignPrivateIP(ip net.IP, node *corev1.Node) error {
+	if err != nil {
+		return err
+	}
+	// 3) Then, add the IP address to the port's allowed_address_pairs.
+	//    TODO: use a more elegant retry mechanism.
+	if err = o.allowIPAddressOnNeutronPort(matchingPort.ID, ip); err != nil && !errors.Is(err, AlreadyExistingIPError) {
+		// Try to clean up the allocated port if adding the IP to allowed_address_pairs failed.
+		// Try this 10 times, but if this operation fails more than that, then user intervention is needed or
+		// the upper layer must call ReleasePrivateIP (because if the neutron port exists and holds
+		// a reservation, then the assign step will not continue after step 2).
+		var errRelease error
+		var releaseStatus string
+		for i := 0; i < 10; i++ {
+			errRelease = o.releaseNeutronIPAddress(*unboundPort, serverID)
+			// If the release operation was successful, then we are done.
+			if errRelease == nil {
+				releaseStatus = "Released neutron port reservation."
+				break
 			}
-			// 3) Then, add the IP address to the port's allowed_address_pairs.
-			//    TODO: use a more elegant retry mechanism.
-			if err = o.allowIPAddressOnNeutronPort(serverPort.ID, ip); err != nil && !errors.Is(err, AlreadyExistingIPError) {
-				// Try to clean up the allocated port if adding the IP to allowed_address_pairs failed.
-				// Try this 10 times, but if this operation fails more than that, then user intervention is needed or
-				// the upper layer must call ReleasePrivateIP (because if the neutron port exists and holds
-				// a reservation, then the assign step will not continue after step 2).
-				var errRelease error
-				var releaseStatus string
-				for i := 0; i < 10; i++ {
-					errRelease = o.releaseNeutronIPAddress(*unboundPort, serverID)
-					// If the release operation was successful, then we are done.
-					if errRelease == nil {
-						releaseStatus = "Released neutron port reservation."
-						break
-					}
-					// Otherwise store the error message and retry.
-					releaseStatus = fmt.Sprintf("Could not release neutron port reservation after %d tries, err: %q", i+1, errRelease)
-				}
-				return fmt.Errorf("could not allow IP address %s on port %s, err: %q. %s", ip.String(), serverPort.ID, err, releaseStatus)
+			// Otherwise store the error message and retry.
+			releaseStatus = fmt.Sprintf("Could not release neutron port reservation after %d tries, err: %q", i+1, errRelease)
+		}
+		return fmt.Errorf("could not allow IP address %s on port %s, err: %q. %s", ip.String(), matchingPort.ID, err, releaseStatus)
+	}
+	// 4) Return nil to indicate success if steps 2 and 3 passed.
+	return nil
+}
+
+// MovePrivateIP implements moving the IP from one node to another to serve cases like a failover.
+// What's different from calling ReleasePrivateIP followed by AssignPrivateIP is that the reservation
+// Neutron port will not get deleted - MovePrivateIP will only change the allowed_address_pairs on the node's
+// ports to remove IP from nodeToDel and add it to nodeToAdd.
```

Comment:
Add something like: "Additionally, MovePrivateIP will recreate a missing reservation port (this is a corner case and should not happen in normal operation)"